In [1]:
import warnings
warnings.filterwarnings('ignore')
import numpy as np
import pandas as pd
import seaborn as sns
from matplotlib import pyplot as plt
%config InlineBackend.figure_format = 'retina'
%matplotlib inline

from  sklearn.tree  import  DecisionTreeClassifier 

# Задача:
# Мы исследуем следующий сценарий: человек решает стэп, не может получить правильный ответ и больше не возвращается к практическим задачам. Что это за шаг такой, который отпугнул максимальное число пользователей?

In [4]:
df1 = pd.read_csv('C:/Program Files/Python37/1/CSV/stepik/event_data_train.csv')
df2 = pd.read_csv('C:/Program Files/Python37/1/CSV/stepik/submissions_data_train.csv')

In [5]:
df1

step_id   timestamp      action  user_id
0          32815  1434340848      viewed    17632
1          32815  1434340848      passed    17632
2          32815  1434340848  discovered    17632
3          32811  1434340895  discovered    17632
4          32811  1434340895      viewed    17632
...          ...         ...         ...      ...
3480698    33375  1526772685      viewed    11677
3480699    32815  1526772750      viewed     5198
3480700    32815  1526772750  discovered     5198
3480701    32815  1526772750      passed     5198
3480702    33892  1526772811      viewed    18526

[3480703 rows x 4 columns]

In [6]:
df2

step_id   timestamp submission_status  user_id
0         31971  1434349275           correct    15853
1         31972  1434348300           correct    15853
2         31972  1478852149             wrong    15853
3         31972  1478852164           correct    15853
4         31976  1434348123             wrong    15853
...         ...         ...               ...      ...
509099   120745  1501946959             wrong     2615
509100   120745  1503059927           correct    13177
509101   120745  1506348153             wrong    15253
509102   120745  1523035316             wrong     1424
509103   120745  1523382825             wrong     9422

[509104 rows x 4 columns]

In [7]:
df1['date'] = pd.to_datetime(df1.timestamp, unit='s')
df1['day'] = df1.date.dt.date

df2['date'] = pd.to_datetime(df2.timestamp, unit='s')
df2['day'] = df2.date.dt.date

In [8]:
df1

step_id   timestamp      action  user_id                date  \
0          32815  1434340848      viewed    17632 2015-06-15 04:00:48   
1          32815  1434340848      passed    17632 2015-06-15 04:00:48   
2          32815  1434340848  discovered    17632 2015-06-15 04:00:48   
3          32811  1434340895  discovered    17632 2015-06-15 04:01:35   
4          32811  1434340895      viewed    17632 2015-06-15 04:01:35   
...          ...         ...         ...      ...                 ...   
3480698    33375  1526772685      viewed    11677 2018-05-19 23:31:25   
3480699    32815  1526772750      viewed     5198 2018-05-19 23:32:30   
3480700    32815  1526772750  discovered     5198 2018-05-19 23:32:30   
3480701    32815  1526772750      passed     5198 2018-05-19 23:32:30   
3480702    33892  1526772811      viewed    18526 2018-05-19 23:33:31   

                day  
0        2015-06-15  
1        2015-06-15  
2        2015-06-15  
3        2015-06-15  
4        2015-06-15  
...             ...  
3480698  2018-05-19  
3480699  2018-05-19  
3480700  2018-05-19  
3480701  2018-05-19  
3480702  2018-05-19  

[3480703 rows x 6 columns]

In [14]:
df2

step_id   timestamp submission_status  user_id                date  \
0         31971  1434349275           correct    15853 2015-06-15 06:21:15   
1         31972  1434348300           correct    15853 2015-06-15 06:05:00   
2         31972  1478852149             wrong    15853 2016-11-11 08:15:49   
3         31972  1478852164           correct    15853 2016-11-11 08:16:04   
4         31976  1434348123             wrong    15853 2015-06-15 06:02:03   
...         ...         ...               ...      ...                 ...   
509099   120745  1501946959             wrong     2615 2017-08-05 15:29:19   
509100   120745  1503059927           correct    13177 2017-08-18 12:38:47   
509101   120745  1506348153             wrong    15253 2017-09-25 14:02:33   
509102   120745  1523035316             wrong     1424 2018-04-06 17:21:56   
509103   120745  1523382825             wrong     9422 2018-04-10 17:53:45   

               day  
0       2015-06-15  
1       2015-06-15  
2       2016-11-11  
3       2016-11-11  
4       2015-06-15  
...            ...  
509099  2017-08-05  
509100  2017-08-18  
509101  2017-09-25  
509102  2018-04-06  
509103  2018-04-10  

[509104 rows x 6 columns]

In [20]:
#[df2[df2['submission_status'] == 'wrong'] & df2.step_id].value_counts()
a = df2.groupby(['step_id', 'submission_status']).count()
a.head(20)

#print(titanic_df.groupby(['Sex', 'Survived'])['PassengerID'].count())
#weekday_counts = berri_bikes.groupby('weekday').aggregate(sum)

timestamp  user_id   date    day
step_id submission_status                                  
31971   correct                 6257     6257   6257   6257
        wrong                   6893     6893   6893   6893
31972   correct                 6434     6434   6434   6434
        wrong                   8960     8960   8960   8960
31976   correct                 6580     6580   6580   6580
        wrong                   5937     5937   5937   5937
31977   correct                 6800     6800   6800   6800
        wrong                   6457     6457   6457   6457
31978   correct                 6094     6094   6094   6094
        wrong                  16084    16084  16084  16084
31979   correct                 3918     3918   3918   3918
        wrong                   5851     5851   5851   5851
31981   correct                 3404     3404   3404   3404
        wrong                   6818     6818   6818   6818
31983   correct                 4017     4017   4017   4017
        wrong                   3410     3410   3410   3410
31986   correct                 3415     3415   3415   3415
        wrong                   1072     1072   1072   1072
31988   correct                 3375     3375   3375   3375
        wrong                    725      725    725    725

Из этого следует, что у step_id(31978) больше всего не правильных ответов wrong(16084).
Ответ 31978

In [21]:
df2.step_id.value_counts()

31978    22178
32031    18221
32202    16950
31972    15394
32075    14534
         ...  
33874     2319
32763     2301
33984     2286
33681     2280
33890     2226
Name: step_id, Length: 76, dtype: int64

In [26]:
list_1 = [1,2,3,4,5]
list_2 = [6,7,8]
list_1[2:2] = list_2

print(list_1)

[1, 2, 6, 7, 8, 3, 4, 5]
